### PR TITLE
Tokenize floats

### PR DESCRIPTION
**Pull request by manuranga**
Part of #262

### CODE REVIEW

File: compiler/modules/front/tests/tokenTests.bal [hunk 1]
```diff
@@ -6,12 +6,23 @@ function test1() {
...
+
+    test:assertEquals(tokenization("0.0 0.0f 0e01 .00"), [[DECIMAL_FP_NUMBER, "0.0"], [DECIMAL_FP_NUMBER, 0.0], [DECIMAL_FP_NUMBER, "0e01"], [DECIMAL_FP_NUMBER, ".00"]]);
+    test:assertEquals(tokenization("-4f 2.0"), ["-", [DECIMAL_FP_NUMBER, 4.0], [DECIMAL_FP_NUMBER, "2.0"]]);
+    test:assertEquals(tokenization("1.2.floor()"), [[DECIMAL_FP_NUMBER, "1.2"], ".", [IDENTIFIER, "floor"], "(", ")"]);
+    test:assertEquals(tokenization("1e-5 1e-5f"), [[DECIMAL_FP_NUMBER, "1e-5"], [DECIMAL_FP_NUMBER, 1.0E-5]]);
+    test:assertEquals(tokenization("1.e1 1.e1F"), [[DECIMAL_FP_NUMBER, "1.e1"], [DECIMAL_FP_NUMBER, 10.0]]);
+    test:assertEquals(tokenization(".5e10 .5e10f"), [[DECIMAL_FP_NUMBER, ".5e10"], [DECIMAL_FP_NUMBER, 5.0E9]]);
+    test:assertEquals(tokenization("-10.0e-2 -10.0e-2f"), ["-", [DECIMAL_FP_NUMBER, "10.0e-2"], "-", [DECIMAL_FP_NUMBER, 0.1]]);
+
+    test:assertEquals(tokenization(".e1"), [".", [IDENTIFIER, "e1"]]);
```
Comment (manuranga) [inline compiler/modules/front/tests/tokenTests.bal:26]: Currently  `1.e-()` gives an error when torkenizing, perhaps it's more correct to torkenize it as `[[DECIMAL_FP_NUMBER, 1.0], [IDENTIFIER, "e"], "-", "(", ")"]]`. That may require significant change.

Comment (jclark) [reply compiler/modules/front/tests/tokenTests.bal:26]: Yes, that is the correct answer (except `[DECIMAL_FP_NUMBER, "1."]`.

We need to be able to do:

`1. e = 1.0`

**Comment by jclark** [comment]
There seems to be a lot of breakage from allowing trailing `.` in a floating point number. We have maybe got do something about that at the spec level. So can we simply not allow that for now?

I have made a spec issue for this. https://github.com/ballerina-platform/ballerina-spec/issues/905

File: compiler/modules/front/token.bal [hunk 2]
```diff
@@ -282,6 +304,80 @@ class Tokenizer {
         return self.err("invalid token");
     }
 
+    private function nextDecimalFPNumber(string preDot) returns Token|err:Syntax {
```
Comment (jclark) [inline compiler/modules/front/token.bal]: Ballerina convention in a case like this is  `Fp` not `FP`

File: compiler/modules/front/token.bal [hunk 3]
```diff
@@ -282,6 +304,80 @@ class Tokenizer {
...
+            }
+            else {
+                self.ungetc(ch);
+                break;
+            }
+        }
+        return [DECIMAL_FP_NUMBER, digits];
+    }
+
+    private function nextDecimalExpedFPNumber(string preExp, Char e) returns Token|err:Syntax {
```
Comment (jclark) [inline compiler/modules/front/token.bal]: No idea what `exped` means

**Review by jclark — CHANGES_REQUESTED** [review]
If we don't accept trailing `.`, we can get rid of all the test case changes.

File: compiler/modules/front/token.bal (hunk 3 above)
Comment (manuranga) [reply compiler/modules/front/token.bal]: I meant to say "Exp-ed" meaning "Exponentiated", I'll go with "nextDecimalExpFpNumber" instead.

File: compiler/modules/front/token.bal [hunk 4]
```diff
@@ -7,9 +7,10 @@ const DECIMAL_NUMBER = 1;
 const STRING_LITERAL = 2;
 const BOOLEAN_LITERAL = 3;
 const HEX_INT_LITERAL = 4;
+const DECIMAL_FP_NUMBER = 5;
 
 // Use string for DECIMAL_NUMBER so we don't get overflow on -int:MAX_VALUE
-type VariableLengthToken [IDENTIFIER, string]|[DECIMAL_NUMBER, string]|[STRING_LITERAL, string]|[HEX_INT_LITERAL, string];
+type VariableLengthToken [IDENTIFIER, string]|[DECIMAL_NUMBER, string]|[STRING_LITERAL, string]|[HEX_INT_LITERAL, string]|[DECIMAL_FP_NUMBER, float|string];
```
Comment (jclark) [inline compiler/modules/front/token.bal]: On further reflection I think a cleaner way to represent the token is: `[DECIMAL_FP_NUMBER, string, FpTypeSuffix?]` where

```
type FpTypeSuffix "f";
```

**Review by jclark — CHANGES_REQUESTED** [review]

**Pull request merged**